### DISCUSSION TITLE
Reanimated 3 Stable Release Plan

### DISCUSSION THREAD

**Discussion by pavangoyal42** [software-mansion/react-native-reanimated#4021] [Q&A]
Hi @tomekzaw, we are using react-native-reanimated in our project. We are facing a burning crash in our Dream11 app on reanimated version **2.9.1** . We are able to repro(15/100 times) the crash with the automation script. Crash can be reproduced with latest stable 2.x.x version as well but not reproducible in **3.0.0.rc.10**.  We can't use RC version with production app. I can't see any update on this thread #3529 after Nov 2022. Please let us know when you guys are planning to push out 3.0.0 stable release ? 
We can help in fixing any open issues or any help we can do to push out the stable release. 

`Fatal Exception: com.facebook.jni.CppException: bad_weak_ptr
       at com.swmansion.reanimated.Scheduler.triggerUI(SourceFile)
       at com.swmansion.reanimated.Scheduler$1.run(SourceFile:24)
       at com.swmansion.reanimated.Scheduler$2.runGuarded(SourceFile:43)
       at com.facebook.react.bridge.GuardedRunnable.run(SourceFile:31)
       at android.os.Handler.handleCallback(Handler.java:883)
       at android.os.Handler.dispatchMessage(Handler.java:100)
       at android.os.Looper.loop(Looper.java:238)
       at android.app.ActivityThread.main(ActivityThread.java:7829)
       at java.lang.reflect.Method.invoke(Method.java)
       at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:512)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:986)`

**Comment by tomekzaw** [comment]
Hi @pavangoyal42, thanks for starting this discussion.

First, let me explain why the error happens on v2 but doesn't on v3. Reanimated 3.0.0-rc.8 introduces new implementation of shared values (#3722). The motivation of the rewrite was to eliminate memory-related issues, such as the one that you are experiencing, which were really difficult to reproduce, and even more difficult to investigate and fix. First, we were considering backporting this change to v2 branch, but the process would be time-consuming and error-prone, so we decided to keep it only for v3 (similarly as with the layout animations rewrite done in https://github.com/software-mansion/react-native-reanimated/pull/3332) and leave v2 with the old battle-tested implementation.

Now let me explain why v3 is still marked as release candidate. In October 2021, we started adding Fabric support to Reanimated. The library integrates quite closely with React Native internals and it took a lot of effort to integrate v2 API with Fabric. We also made a decision not to bring Fabric support to legacy v1 API because we thought that hardly anyone will use deprecated API on the new experimental architecture. Given that, the only reason why we released v3 as RC and not stable was the fact that Layout Animations were not yet implemented on Fabric. The plan was to release what we already had as an RC version so that the community could start using it earlier and provide feedback while we would have some more time to work on the missing feature. However, priorities have changed as we had to focus on other maintenance tasks such as adding support for next versions of React Native, improving build config, or fixing long-standing issues related to memory leaks or crashes.

Finally, let me explain why we postpone the release of v3 stable for so long. Officially, Reanimated 2 supports React Native 0.62+ and there's a lot of configurations that we need to test thoroughly before each release. Bumping the major version lets us not only to introduce some API-wise changes (like removing v1 API) but also finally drop the support for older versions of React Native (currently we are thinking of support either 0.65+ or 3 latest minor releases, i.e. 0.69+). Another reason is that we also wanted to gather some feedback on shared value and layout animation rewrites. Finally, we would like to include two more features: shared element transitions (#3827) and inline styles (#3354), which hopefully should encourage the community to migrate to v3.

> any help we can do to push out the stable release.

Actually, it would be great if you could upgrade to 3.0.0-rc.10 and tell us if it works as expected in your app. We know that v3 is already used in a number of production apps but your feedback would give us even more confidence to finally release 3.0.0-stable.

**Reply by pavangoyal42** [reply]
Hi @tomekzaw Thanks for the revert. We are planning ti integrate this version in our app. Will check and get back. We have one more query. We are seeing ANR as well related to reanimated #3879. Is it also solved in the 3.0.0-rc.10 ?

**Reply by tomekzaw** [reply]
Hi @pavangoyal42, unfortunately #3879 is a long-standing issue that we still need to find a solution for.

**Comment by astanb** [comment]
We are experiencing lots of memory related problems - app crashes when it gets full and also when we restart after a code push if certain shared objects are in memory - we'd love to try out V3 but unfortunately we also rely heavily on vision camera which will not work with V3, so we are stuck. We also use other libraries that rely on reanimated so I assume they would also need updating before we could use V3.

**Reply by tomekzaw** [reply]
> we also rely heavily on vision camera which will not work with V3, so we are stuck

cc @mrousavy, see https://github.com/mrousavy/react-native-vision-camera/issues/1376

> We also use other libraries that rely on reanimated so I assume they would also need updating before we could use V3.

I believe most of the libraries that rely on Reanimated actually use v2 API which is perfectly compatible with v3. According to [this poll](https://twitter.com/kzzzf/status/1605917602735214592), most users refrain from upgrading only because v3 is still marked as rc, and not because of lack of compatibility. @alexstanbury Can you please check and let us know if any of your dependencies uses v1 API? We are considering devoting some of our time to migrating such libraries to the V2 API to allow the community to migrate to Reanimated v3.

**Reply by astanb** [reply]
Hi, vision camera doesn't work with V3, see https://github.com/software-mansion/react-native-reanimated/issues/3831#issuecomment-1359263674

**Reply by ucheNkadiCode** [reply]
Also [react-native-draggable-flatlist](https://github.com/computerjazz/react-native-draggable-flatlist#readme) doesn't support v3 either and uses V1 functions. It relies on plenty of v1 paradigmss such as 
-  `outerScrollOffset?: Animated.Node<number>;`
- `Animated.Code`
- `Animated.SpringConfig`
-   clockRunning, not, startClock, stopClock,
- set, cond, add, sub, block, eq
- useValue(_numberVar)
-
I wanted to refactor the library myself since @computerjazz seems busy, but I have no clue where to start with refactoring all of these types and variables. 

I'd love a guide to help make it clear which aspects of V1 went away for good and how to translate something like Animated.Code into its new counterpart. 

@tomekzaw @piaskowyk (since y'all are super helpful and responsive1"

**Reply by computerjazz** [reply]
@ucheNkadiCode sounds like you're on an old version of react native draggable flatlist, it's supported reanimated v3 since v4.0.0 was released last year.

**Reply by ucheNkadiCode** [reply]
You were totally right, fortunately I just needed to update react navigation/drawer and draggable flat list to the very latest and it was supported

**Comment by pavangoyal42** [comment]
@tomekzaw I can confirm after 3.0.0. we are not seeing any crash.

**Reply by yepMad** [reply]
Are you using shared-elements? Do it work in release build?

**Reply by wcastand** [reply]
as far as i know i'm waiting for this https://github.com/software-mansion/react-native-reanimated/issues/4135

it was merged but not published so still crashing on Android 7/8 in theory unless i missed something.

**Comment by gnair03** [comment]
@tomekzaw 
Pls refer to this issue for context - [#37126](https://github.com/facebook/react-native/issues/37126)

As a plan to fix this, we have developed our own custom sticky behavior on top of Reanimated's Scrollview component. On 3.0.0-rc10, we were experiencing a jumpy behavior in this custom sticky implementation on fast scroll. This jumpiness is NOT observed on 3.1.0, and hence we are now migrating to 3.1.0.
But on migration, the app size has increased by 1.5 MB. Can you help me with the possible reasons behind this considerable size increase? Is there anyway this delta can be reduced? 

Pls let me know
